content/blog/2022-04-07-compiling-black-with-mypyc-part-3.md

```diff
@@ -0,0 +1,185 @@
+tool used to [provide direct][diff-shades-comment-1]
+[feedback on PRs][diff-shades-comment-2]. I tried to integrate it into the wheel build
+workflow as part of the test step, but that turned out be very painful and I backed out of
+it since I had a stable release to manage and publish!
+
+*The TL;DR of it is as follows:*
+
+Building Black with mypyc usually requires disabling pip's build isolation to work
+properly as mypyc is not a standard build dependency of Black. This messes up the
+installation of diff-shades which is packaged using flit. So I hacked up a script to edit
+the `[build-system].requires` field in `pyproject.toml` to include mypyc pre-build, but
+the isolation somehow broke the linker search path or something.
+
+```text
+clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -g0 -fPIC -I/tmp/pip-build-env-9bfrmy6j/overlay/lib/python3.7/site-packages/mypyc/lib-rt -Ibuild -I/opt/python/cp37-cp37m/include/python3.7m -c build/__native_f2d4935fd652bc9ef29d.c -o build/temp.linux-x86_64-3.7/build/__native_f2d4935fd652bc9ef29d.o -O3 -Werror -Wno-unused-function -Wno-unused-label -Wno-unreachable-code -Wno-unused-variable -Wno-unused-command-line-argument -Wno-unknown-warning-option
+    clang -shared -g0 build/temp.linux-x86_64-3.7/build/__native_f2d4935fd652bc9ef29d.o -o build/lib.linux-x86_64-3.7/f2d4935fd652bc9ef29d__mypyc.cpython-37m-x86_64-linux-gnu.so
+    /usr/bin/ld: cannot find crtbeginS.o: No such file or directory
+    /usr/bin/ld: cannot find -lgcc
+    /usr/bin/ld: cannot find -lgcc_s
+    clang: error: linker command failed with exit code 1 (use -v to see invocation)
+    error: command '/usr/bin/clang' failed with exit code 1
+```
+
+I honestly still have no idea what's wrong.
+
+Anyway since I was like an hour or more into build errors, I just decided to call it done
+and fall back to the basic testing that was already working.[^12] I triggered the last
+workflow run of the day, downloaded the artifacts, tested one of them locally to make sure
+nothing was on fire, and pushed 'em to PyPI and that's how release 22.1.0 was born 🎉.
+
+The core team chat was quite lively for the next hour to say the least, after all we just
+finished a major milestone! Also yeah, we *may* have done a lot of publicizing for this
+release which is why it was everywhere for a while :wink:
+
+### Post release calm
+
+To be honest I haven't really seen anyone comment on the fact Black is now compiled with
+mypyc yet short of this one [mini blog post][that-one-post-about-black-mypyc] ... which is
+both disappointing since I spent so much time on this effort, but also calming since
+usually buzz is sparked by bugs and whatnot.
+
+Having mentioned bugs and crashes, so far we've received two reports,
+[one bug probably from mypyc and an unintentional restriction on Black's unofficial APIs][not-too-many-fires-so-far].
+
+Although version 22.1.0 still unintentionally broke quite a few integrations, though they
```

Comment:
```suggestion
Version 22.1.0 still unintentionally broke quite a few integrations, though they
```

Reply:
I'm really bad at remembering to check the whole sentence when I change one clause of it. I reworded the second clause and totally forgot to check it still made sense with the first \*sigh*